### ISSUE TITLE
Use tab together with luasnip

### ISSUE THREAD

**Issue by gegoune** [hrsh7th/nvim-cmp#21]
With compe I had `{s-,}tab` mapped as such

```lua
local check_back_space = function()
  local col = vim.fn.col '.' - 1
  return col == 0 or vim.fn.getline('.'):sub(col, col):match '%s' ~= nil
end

_G.tab_complete = function()
  if vim.fn.pumvisible() == 1 then
    return t '<C-n>'
  elseif luasnip.expand_or_jumpable() then
    return t '<Plug>luasnip-expand-or-jump'
  elseif check_back_space() then
    return t '<Tab>'
  else
    return vim.fn['compe#complete']()
  end
end

_G.s_tab_complete = function()
  if vim.fn.pumvisible() == 1 then
    return t '<C-p>'
  elseif luasnip.jumpable(-1) then
    return t '<Plug>luasnip-jump-prev'
  else
    return t '<S-Tab>'
  end
end

map({ 'i', 's' }, '<Tab>', 'v:lua.tab_complete()', { expr = true, noremap = false })
map({ 'i', 's' }, '<S-Tab>', 'v:lua.s_tab_complete()', { expr = true, noremap = false })
```
so depending if pum was visible it would jump between placeholders or navigate menu.

How can I replicate it with cmp, currently having: 

```lua
  mapping = {
    ['<S-Tab>'] = cmp.mapping.prev_item(),
    ['<Tab>'] = cmp.mapping.next_item(),
}
```
?

Thanks!

**Comment by hrsh7th** [comment]
I'm considering the API but now you can use the following config.

```lua
      ['<C-n>'] = function(core, fallback)
        if vim.fn.pumvisible() == 1 then
          vim.fn.feedkeys(vim.api.nvim_replace_termcodes('<C-n>', true, true, true), 'n')
        elseif luasnip.expand_or_jumpable() then
          vim.fn.feedkeys(vim.api.nvim_replace_termcodes('<Plug>luasnip-expand-or-jump', true, true, true), '')
        elseif not check_backspace() then
          vim.fn.feedkeys(vim.api.nvim_replace_termcodes('<Tab>', true, true, true), 'n')
        else
          cmp.mapping.complete()(core, fallback)
        end
      end,
      ['<C-p>'] = function(core, fallback)
        if vim.fn.pumvisible() == 1 then
          vim.fn.feedkeys(vim.api.nvim_replace_termcodes('<C-p>', true, true, true), 'n')
        elseif luasnip.jumpable(-1) then
          vim.fn.feedkeys(vim.api.nvim_replace_termcodes('<Plug>luasnip-jump-prev', true, true, true), '')
        else
          vim.fn.feedkeys(vim.api.nvim_replace_termcodes('<S-Tab>', true, true, true), 'n')
        end
      end,
```

**Comment by windwp** [comment]
i use this trick to mapping complete ` return vim.fn['compe#complete']()` replace with ` return _G.cmp.utils.keymap.expr('<CR>')` :)

**Comment by gegoune** [comment]
Thank you @hrsh7th. I have tried that, Tab works, but going backwards with S-Tab doesn't. It just inserts two spaces snippet like 

```lua
function <placeholder1>(<placeholder2>)
  <placceholder3>
end
```

I can jump between placeholders 1 and 2 back and forth, but when I get to 3rd one I cannot come back, tab inserts spaces (per my config I think).

My mapping says:
```
i  <S-Tab>     *@v:lua.cmp.utils.keymap.expr("<S-Tab>")
```

Any ideas where to look next?

**Comment by hrsh7th** [comment]
@gegoune I tried it with luasnip and I found the cause maybe. The `luasnip.jumpable(-1)` returns `false` in this case.

**Comment by hrsh7th** [comment]
I've improved (a bit) the above example config

**Comment by leiserfg** [comment]
@gegoune it will be useful if you share the real snippet instead of a pseudocode .

**Comment by hrsh7th** [comment]
@gegoune I've asked luasnip dev teams about it. They said it is expected behavior.
The LuaSnip treats `$0` is snippet ending so if the cursor was jumped to `$0`, the snippet will deactivated.

**Comment by hrsh7th** [comment]
I think it can be closed. @gegoune

**Comment by L3MON4D3** [comment]
> Thank you @hrsh7th. I have tried that, Tab works, but going backwards with S-Tab doesn't. It just inserts two spaces snippet like
> 
> ```lua
> function <placeholder1>(<placeholder2>)
>   <placceholder3>
> end
> ```
> 
> I can jump between placeholders 1 and 2 back and forth, but when I get to 3rd one I cannot come back, tab inserts spaces (per my config I think).


If you'd like to still be able to jump back from the `i(0)`, set `history` to `true`:
```lua
require'luasnip'.config.setup({history=true})
```

**Comment by gegoune** [comment]
Aha, I see. Thanks you all! I think my confusion originated from getting different snippets for `function` actually. One snippet would be `fu`, the other `function`

<img width="138" alt="Screenshot 2021-08-16 at 10 58 45" src="https://user-images.githubusercontent.com/69750637/129538310-4dfa4dfe-2ea7-494c-9613-d45b5d5f5a27.png">

<img width="269" alt="Screenshot 2021-08-16 at 11 01 01" src="https://user-images.githubusercontent.com/69750637/129538581-85b5c2d3-bd09-4078-88ea-11b5bc19453b.png">

Now, in compe I can actually go back and forth in both snippets, even back from the body of the function. Can't go back from body in cmp. Difference is coming from how compe and cmp are both setting up mappings I think. Highlighted placeholders (latest changes to luasnip?) put you in visual (select?) mode. Seems I had `s` mode mappings with compe:

```lua
map({ 'i', 's' }, '<Tab>', 'v:lua.tab_complete()', { expr = true, noremap = false })
map({ 'i', 's' }, '<S-Tab>', 'v:lua.s_tab_complete()', { expr = true, noremap = false })
```

So basically getting placeholder highlighted put me in mode in which I no longer have my `{s-,}tab` mappings for cmp.

@hrsh7th could you api for setting up keymaps take modes as an extra argument?

**Comment by hrsh7th** [comment]
It's insightful. I have to consider it.

**Comment by hrsh7th** [comment]
@gegoune I've added the API for it (But it is not stable now. Sorry)

Could you try it?
https://github.com/hrsh7th/nvim-cmp#mapping-type-tablestring-funcore-cmpcore-fallback-function

**Comment by gegoune** [comment]
Super lovely, thank you very much @hrsh7th, super star! Works very nice, as expected and what I am used to from compe.

**Comment by rhcher** [comment]
Sorry for leaving messages in this issue, but I want to know how to implement this: I use tab to confirm item when pumvisible() == 1 and jump next when vsnip available == 1 and else just TAB. @hrsh7th

**Comment by hrsh7th** [comment]
Try with this configuration.
```
      ['<Tab>'] = function(core, fallback)
        if vim.fn.pumvisible() == 1 then
          vim.fn.feedkeys(vim.api.nvim_replace_termcodes('<C-n>', true, true, true), 'n')
        elseif vim.fn['vsnip#available']() == 1 then
          vim.fn.feedkeys(vim.api.nvim_replace_termcodes('<Plug>(vsnip-expand-or-jump)', true, true, true), '')
        elseif not check_backspace() then
          cmp.mapping.complete()(core, fallback)
        else
          fallback()
        end
      end,
      ['<S-Tab>'] = function(core, fallback)
        if vim.fn.pumvisible() == 1 then
          vim.fn.feedkeys(vim.api.nvim_replace_termcodes('<C-p>', true, true, true), 'n')
        elseif vim.fn['vsnip#jumpable'](-1) == 1 then
          vim.fn.feedkeys(vim.api.nvim_replace_termcodes('<Plug>(vsnip-jump-prev)', true, true, true), '')
        else
          fallback()
        end
      end,
```

**Comment by rhcher** [comment]
I'm more used to using \<C-n> to navigate, and what I want in cmp is actually that🤣:
``` lua
local t = function(str)
  return vim.api.nvim_replace_termcodes(str, true, true, true)
end

_G.my_tab_complete = function()
  if vim.fn.pumvisible() == 1 then
    return vim.fn["compe#confirm"]()
  elseif vim.fn.call("vsnip#available", { 1 }) == 1 then
    return t "<Plug>(vsnip-expand-or-jump)"
  else
    return t "<Tab>"
  end
end
```

**Comment by hrsh7th** [comment]
Do your problem was solved by the above config?

**Comment by rhcher** [comment]
No, beacuse it's the solution when I use compe.

**Comment by hrsh7th** [comment]
Sorry for my poor english...

Do the following config isn't help you?
```
      ['<Tab>'] = function(core, fallback)
        if vim.fn.pumvisible() == 1 then
          vim.fn.feedkeys(vim.api.nvim_replace_termcodes('<C-n>', true, true, true), 'n')
        elseif vim.fn['vsnip#available']() == 1 then
          vim.fn.feedkeys(vim.api.nvim_replace_termcodes('<Plug>(vsnip-expand-or-jump)', true, true, true), '')
        elseif not check_backspace() then
          cmp.mapping.complete()(core, fallback)
        else
          fallback()
        end
      end,
      ['<S-Tab>'] = function(core, fallback)
        if vim.fn.pumvisible() == 1 then
          vim.fn.feedkeys(vim.api.nvim_replace_termcodes('<C-p>', true, true, true), 'n')
        elseif vim.fn['vsnip#jumpable'](-1) == 1 then
          vim.fn.feedkeys(vim.api.nvim_replace_termcodes('<Plug>(vsnip-jump-prev)', true, true, true), '')
        else
          fallback()
        end
      end,
      ```

**Comment by rhcher** [comment]
As I understand it, this configuration means that the tab key is equal to \<C-n> when pumvisible == 1. If I understand it correctly, then this is not what I want, I want the tab key to behave is this when pumvisible == 1
``` lua
      ['<tab>'] = cmp.mapping.confirm({
        behavior = cmp.ConfirmBehavior.Replace,
        select = true,
      })
```

**Comment by hrsh7th** [comment]
I've understand maybe. I will paste a sample config.

**Comment by hrsh7th** [comment]
Try it.
```lua
{
  ['<CR>'] = cmp.mapping.confirm({
    select = true,
    behavior = cmp.ConfirmBehavior.Replace,
  }),
  ['<Tab>'] = function(core, fallback)
    if vim.fn.pumvisible() == 1 then
      vim.fn.feedkeys(vim.api.nvim_replace_termcodes('<CR>', true, true, true), '')
    elseif vim.fn['vsnip#available']() == 1 then
      vim.fn.feedkeys(vim.api.nvim_replace_termcodes('<Plug>(vsnip-expand-or-jump)', true, true, true), '')
    else
      fallback()
    end
  end,
}
```

**Comment by rhcher** [comment]
This is exactly what I wanted, but since I didn't want to pollute the \<tab> key (it has other tasks), I replaced it with a strange enough key. Thank you for your great work.

**Comment by gegoune** [comment]
@hrsh7th What's the correct way of mapping for multiple modes now?

Tried this:

```lua
  mapping = {
    ['<Tab>'] = cmp.mapping(function(fallback)
      if vim.fn.pumvisible() == 1 then
        vim.fn.feedkeys(t '<C-n>', 'n')
      elseif check_backspace() then
        vim.fn.feedkeys(t '<Tab>', 'n')
      elseif luasnip.expand_or_jumpable() then
        vim.fn.feedkeys(t '<Plug>luasnip-expand-or-jump', '')
      else
        fallback()
      end
    end, {
      'i',
      's',
    }),
    ['<S-Tab>'] = cmp.mapping(function(fallback)
      if vim.fn.pumvisible() == 1 then
        vim.fn.feedkeys(t '<C-p>', 'n')
      elseif luasnip.jumpable(-1) then
        vim.fn.feedkeys(t '<Plug>luasnip-jump-prev', '')
      else
        fallback()
      end
    end, {
      'i',
      's',
    }),
    ['<CR>'] = cmp.mapping.confirm {
      behavior = cmp.ConfirmBehavior.Replace,
      select = true,
    },
  },
```

but tabs don't work, neither does expanding snippets with enter.

**Comment by uga-rosa** [comment]
@gegoune Now this works.
```lua
    ["<tab>"] = function(fallback)
      if vim.fn.pumvisible() == 1 then
        vim.fn.feedkeys(t("<C-n>"), "n")
      elseif luasnip.expand_or_jumpable() then
        vim.fn.feedkeys(t("<Plug>luasnip-expand-or-jump"), "")
      elseif check_back_space() then
        vim.fn.feedkeys(t("<tab>"), "n")
      else
        fallback()
      end
    end,
    ["<S-tab>"] = function(fallback)
      if vim.fn.pumvisible() == 1 then
        vim.fn.feedkeys(t("<C-p>"), "n")
      elseif luasnip.jumpable(-1) then
        vim.fn.feedkeys(t("<Plug>luasnip-jump-prev"), "")
      else
        fallback()
      end
    end,
```
See also the readme.

**Comment by gegoune** [comment]
@uga-rosa No, it doesn't work as expected, misses visual mode mappings.

**Comment by uga-rosa** [comment]
Isn't it because of the position of check_back_space? This worked well.
```lua
    ["<tab>"] = cmp.mapping(function(fallback)
      if vim.fn.pumvisible() == 1 then
        vim.fn.feedkeys(t("<C-n>"), "n")
      elseif luasnip.expand_or_jumpable() then
        vim.fn.feedkeys(t("<Plug>luasnip-expand-or-jump"), "")
      elseif check_back_space() then
        vim.fn.feedkeys(t("<tab>"), "n")
      else
        fallback()
      end
    end, {
      "i",
      "s",
    }),
    ["<S-tab>"] = cmp.mapping(function(fallback)
      if vim.fn.pumvisible() == 1 then
        vim.fn.feedkeys(t("<C-p>"), "n")
      elseif luasnip.jumpable(-1) then
        vim.fn.feedkeys(t("<Plug>luasnip-jump-prev"), "")
      else
        fallback()
      end
    end, {
      "i",
      "s",
    }),
```
I'm not sure why enter doesn't expand the snippet. You have set up a snippet engine, right?

**Comment by gegoune** [comment]
I have, and it works as soon as tabs got fixed, which is puzzling at least. And your tab mappings do work nicely, thanks!

**Comment by kdheepak** [comment]
@hrsh7th @uga-rosa For me this doesn't work consistently:

https://user-images.githubusercontent.com/1813121/130792598-544b2689-503d-445b-87be-2ce70a4ca806.mov

The following is my full configuration:

```lua
local function check_backspace()
  local col = vim.fn.col(".") - 1
  if col == 0 or vim.fn.getline("."):sub(col, col):match("%s") then
    return true
  else
    return false
  end
end

local function T(str)
  return vim.api.nvim_replace_termcodes(str, true, true, true)
end

local cmp = require("cmp")
local luasnip = require("luasnip")
local lspkind = require("lspkind")
local check_backspace = require("kd/utils").check_backspace

cmp.setup({

  -- You should change this example to your chosen snippet engine.
  snippet = {
    expand = function(args)
      return require("luasnip").lsp_expand(args.body)
    end,
  },

  completion = {
    completeopt = "menu,menuone,noselect",
  },

  -- You must set mapping.
  mapping = {
    ["<C-p>"] = cmp.mapping.select_prev_item(),
    ["<C-n>"] = cmp.mapping.select_next_item(),
    ["<C-d>"] = cmp.mapping.scroll_docs(-4),
    ["<C-f>"] = cmp.mapping.scroll_docs(4),
    ["<C-Space>"] = cmp.mapping.complete(),
    ["<C-e>"] = cmp.mapping.close(),
    ["<ESC>"] = cmp.mapping.close(),
    ["<CR>"] = cmp.mapping.confirm({
      behavior = cmp.ConfirmBehavior.Replace,
      select = true,
    }),
    ["<Tab>"] = cmp.mapping(function(fallback)
      if vim.fn.pumvisible() == 1 then
        vim.fn.feedkeys(T("<C-n>"), "n")
      elseif luasnip.expand_or_jumpable() then
        vim.fn.feedkeys(T("<Plug>luasnip-expand-or-jump"), "")
      elseif check_backspace() then
        vim.fn.feedkeys(T("<Tab>"), "n")
      else
        fallback()
      end
    end, {
      "i",
      "s",
    }),
    ["<S-Tab>"] = cmp.mapping(function(fallback)
      if vim.fn.pumvisible() == 1 then
        vim.fn.feedkeys(T("<C-p>"), "n")
      elseif luasnip.jumpable(-1) then
        vim.fn.feedkeys(T("<Plug>luasnip-jump-prev"), "")
      else
        fallback()
      end
    end, {
      "i",
      "s",
    }),
  },

  formatting = {
    format = function(entry, vim_item)
      vim_item.kind = lspkind.presets.default[vim_item.kind] .. " " .. vim_item.kind
      vim_item.menu = ({
        nvim_lsp = "[L]",
        emoji = "[E]",
        path = "[F]",
        calc = "[C]",
        vsnip = "[S]",
        buffer = "[B]",
      })[entry.source.name]
      vim_item.dup = ({
        buffer = 1,
        path = 1,
        nvim_lsp = 0,
      })[entry.source.name] or 0
      return vim_item
    end,
  },

  -- You should specify your *installed* sources.
  sources = {
    { name = "nvim_lsp" },
    { name = "luasnip" },
    { name = "nvim_lua" },
    { name = "emoji" },
    { name = "calc" },
    { name = "path" },
    { name = "latex_symbols" },
    {
      name = "buffer",
      opts = {
        get_bufnrs = function()
          local bufs = {}
          for _, win in ipairs(vim.api.nvim_list_wins()) do
            bufs[vim.api.nvim_win_get_buf(win)] = true
          end
          return vim.tbl_keys(bufs)
        end,
      },
    },
  },
})

```

**Comment by kdheepak** [comment]
I think my problem at least is that the pop up is open some times and isn’t open other times. When it isn’t open, everything works as expected. I haven’t modified the keyword pattern or anything, so I’m not sure why the autocomplete is being triggered in this case.

**Comment by hrsh7th** [comment]
@kdheepak Your key mapping is the correct but the second case seems completion was triggered so You can't insert `<Tab>` as-is but it will be `<C-n>` instead.

You can set the following configuration or update to latest nvim-cmp.

```lua
cmp.setup {
  completion = {
    autocomplete = { cmp.TriggerEvent.TextChanged }
  }
}
```
(This configuration is default by latest nvim-cmp)

**Comment by kdheepak** [comment]
Thanks for the suggestion, I now understand what was going on. I upgraded to the latest. This only _almost_ works. However, now I can't use double tab:

https://user-images.githubusercontent.com/1813121/130805634-00df0444-6cbc-464b-91e8-f58118f54392.mov

I'm expecting two or more tabs to cause issues. This isn't the case with tab in new lines however.

https://user-images.githubusercontent.com/1813121/130806033-5849175e-6cdd-4a5d-b30f-616617612dbb.mov

In new lines, it works the way I'd expect it to. Does that make sense?

**Comment by hrsh7th** [comment]
Because sumneko lua language server specifies space character as trigger characters.
Hm...

**Comment by kdheepak** [comment]
Ah. I didn’t think to try in other languages. I’m sure this is a quandary.

**Comment by uga-rosa** [comment]
Hm? Just to be sure, check `imap <C-n>` for any other mappings.

**Comment by kdheepak** [comment]
Should I open a new issue to track this?

**Comment by kdheepak** [comment]
`:verbose imap <Tab>`:

```
i  <Tab>       *@v:lua.cmp.utils.keymap.expr("i", "<Tab>")
        Last set from anonymous :source```

**Comment by hrsh7th** [comment]
OK. If you allow to disable the auto-completion, the problem can be solved.

**Comment by kdheepak** [comment]
It's not ideal, since I lose discoverability. In my experience when using this kind of set up disabling auto-complete, I basically never ended up using snippets. With the pop up, it is easy to see what available options exist. And sometimes, the pop complete doesn't have what I want, so I'll have to type it in manually anyway. So being able to see the list always is extremely useful.

Disabling auto-completion though is the only way I can currently use this, because I use `<Tab>` to indent way too much :)

**Comment by hrsh7th** [comment]
What's happening with this issue

1. Press `<Tab>` and invoke mapping function.
2. Matches `check_backspace` condition so `<Tab>` will be inserted as-is.
3. The buffer contents will be changed because `<Tab>` is inserted.
4. Auto-completion will be invoked because buffer content is changed (`cmp.TriggerEvent.TextChanged`)
5. The `sumneko_lua` server specifies ` ` as trigger character so completion menu will be shown.

It may be useful for some people to be able to customize the trigger character.

**Comment by uga-rosa** [comment]
As a quick solution, you can also indent with i_CTRL-T (even if it's not at the beginning of the line).

**Comment by cathaysia** [comment]
Here is my solution, it works good for me and as much as simple: 
```lua
['<Tab>'] = function(core, fallback)
    if vim.fn.pumvisible() == 1 then
        vim.fn.feedkeys(vim.api.nvim_replace_termcodes('<C-n>', true, true, true), 'n')
    elseif luasnip.expand_or_jumpable() then
        vim.fn.feedkeys(vim.api.nvim_replace_termcodes('<Plug>luasnip-expand-or-jump', true, true, true), '')
    else
        cmp.mapping.complete()(core, fallback)
    end
end,
```
![Peek 2021-10-03 07-48](https://user-images.githubusercontent.com/42114817/135734639-d58343eb-0767-44d0-b20c-4c047895178f.gif)
